File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/detect/Detect.kt
package jp.co.toukei.log.trustar.detect

import android.location.Location
import android.os.SystemClock
import io.reactivex.rxjava3.core.BackpressureStrategy
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.disposables.Disposable
import jp.co.toukei.log.lib.observeOnIO
import jp.co.toukei.log.lib.util.MapDisposable
import jp.co.toukei.log.lib.util.rxAccelerometerDetectionMillis
import jp.co.toukei.log.lib.util.withLastElementOnCancel
import jp.co.toukei.log.trustar.Current
import jp.co.toukei.log.trustar.db.user.entity.BinHeader
import jp.co.toukei.log.trustar.db.user.entity.SensorDetectEvent
import jp.co.toukei.log.trustar.user.UserInfo
import splitties.systemservices.sensorManager
import java.util.concurrent.TimeUnit
import kotlin.math.max

fun restEvent(
    working: Flowable<List<BinHeader>>,
    location: Flowable<Location>,
    user: UserInfo,
): Flowable<Pair<String, DetectRestEvent>> {
    val restDistanceMeter = user.restDistanceMeter
    val restTimeInMillis = user.restTimeInMin * 60_000L

    return working.eachWorkingBin(
        internalCancelEmit = { _, p ->
            val r = p?.second
            if (r is DetectRestEvent.RestDetected) {
                val now = SystemClock.elapsedRealtimeNanos()
                val d = DetectRestEvent.MoveDetected(r, now)
                val over = d.elapsedRestMillis() > restTimeInMillis
                if (over) p.first to d else null
            } else null
        }
    ) { b ->
        val allocationNo = b.allocationNo
        DetectRestEvent.Reset()
            .detectAfter(
                locationFlow = location,
                restDistance = restDistanceMeter,
                restSpeedMpMs = restDistanceMeter.toDouble() / 15_000,
                restTimeInMillis = restTimeInMillis,
                speedDetectSample = 10_000,
            )
            .map { allocationNo to it }
            .onErrorComplete()
    }.onBackpressureBuffer()
}

fun sensorDetect(
    working: Flowable<List<BinHeader>>,
    binHeaderDelayInMillis: Long,
    workDelayInMillis: Long,
    detectControl: (allocationNo: String) -> Flowable<Boolean>,
): Flowable<SensorRecord> {
    return Flowable.defer {
        val sensorEvent = sensorManager
            .rxAccelerometerDetectionMillis(100, 10, 10_000, 4.3F)
            .observeOnIO()
            .share()
        working.eachWorkingBin(
            internalCancelEmit = { b, _ ->
                val now = System.currentTimeMillis()
                val from = now - binHeaderDelayInMillis
                SensorRecord.DropRecent(b.allocationNo, from, now, true)
            }
        ) { b ->
            detectControl(b.allocationNo)
                .distinctUntilChanged()
                .switchMap {
                    if (it) {
                        sensorEvent
                            .map { eventDate ->
                                SensorRecord.Record(
                                    SensorDetectEvent(b, Current.lastLocation, eventDate)
                                )
                            }
                            .delaySubscription(workDelayInMillis, TimeUnit.MILLISECONDS)
                    } else {
                        val now = System.currentTimeMillis()
                        val from = now - workDelayInMillis
                        Flowable.just(
                            SensorRecord.DropRecent(b.allocationNo, from, now, false)
                        )
                    }
                }
                .delaySubscription(
                    binHeaderDelayInMillis - max(
                        0L,
                        System.currentTimeMillis() - (b.startLocation?.date ?: 0L)
                    ),
                    TimeUnit.MILLISECONDS
                )
                .onErrorComplete()
        }.onBackpressureLatest()
    }
}

fun <T : Any> Flowable<List<BinHeader>>.eachWorkingBin(
    internalCancelEmit: ((BinHeader, lastElement: T?) -> T?)? = null,
    flowable: (BinHeader) -> Flowable<T>,
): Flowable<T> {
    return Flowable.create({ emitter ->
        var onCancelEmit = internalCancelEmit
        val mapDisposable = MapDisposable<String>()
        val d = subscribe { ls ->
            val map = mutableMapOf<String, Disposable>()
            ls.filter { it.binStatus.isWorking() }.forEach { b ->
                val k = b.allocationNo
                val d = mapDisposable.take(k)
                val x = if (d == null || d.isDisposed) {
                    flowable(b)
                        .withLastElementOnCancel {
                            onCancelEmit?.invoke(b, it)?.let(emitter::onNext)
                        }
                        .subscribe(emitter::onNext, emitter::onError)
                } else {
                    d
                }
                map.put(k, x)?.dispose()
            }
            mapDisposable.clear()
            map.forEach {
                mapDisposable.put(it.key, it.value)
            }
        }
        emitter.setCancellable {
            onCancelEmit = null
            d.dispose()
            mapDisposable.dispose()
        }
    }, BackpressureStrategy.MISSING)
}

File: LogiOS/TrustarApp/Detect/Detect.swift
//
//  Detect.swift
//  TrustarApp
//
//  Created by DionSoftware on 2023/03/23.
//  Copyright © 2023 DionSoftware. All rights reserved.
//

import Foundation
import Combine
import CoreMotion
import CoreLocation

func restEvent(
    cancelableBags: inout Set<AnyCancellable>,
    working: AnyPublisher<[BinHeader], Never>,
    location: AnyPublisher<CLLocation, Never>,
    user: UserInfo
) -> AnyPublisher<(String, DetectRestEvent), Never> {
    let restDistanceMeter = user.restDistance
    let restTimeInMillis: Int64 = Int64(user.restTime * 60_000) //miniute
    
    return working.eachWorkingBin(
        cancelableBags: &cancelableBags,
        internalCancelEmit: { _, p in
            let r = p?.1
            if (r is DetectRestEvent.RestDetected) {
                //let now = ProcessInfo.processInfo.systemUptime * 1000 //miliseconds
                let now = Date().milisecondsSince1970
                let rDetected = r as! DetectRestEvent.RestDetected
                let d = DetectRestEvent.MoveDetected(rest: rDetected, endRealtimeNaos: Int64(now))
                let over = d.elapsedRestMillis() > restTimeInMillis
                if (over) {
                    return (p!.0, d)
                } else {
                    return nil
                }
            } else {
                return nil
            }
        },
        flowable: { b in
            //let allocationNo = b.allocationNo
            DetectRestEvent.Reset(elapsedMillis: nil)
                .detectAfter(locationFlow: location,
                             restDistance: restDistanceMeter,
                             restSpeedMpMs: Double(restDistanceMeter) / 15_000,
                             restTimeInMillis: restTimeInMillis,
                             speedDetectSample: 10_000)
                .map { restEvent in
                    return (b.allocationNo, restEvent)
                }.eraseToAnyPublisher()
        }).map({ $0.1 }).eraseToAnyPublisher()
}

func sensorDetect(
    cancelableBags: inout Set<AnyCancellable>,
    working: AnyPublisher<[BinHeader], Never>,
    binHeaderDelayInMillis: Int64,
    workDelayInMillis: Int64,
    sensorManager : CMMotionManager,
    detectControl: @escaping (String) -> AnyPublisher<Bool, Never>
) -> AnyPublisher<SensorRecord, Never> {
    
    let sensorEvent = sensorManager
        .rxAccelerometerDetectionMillis(sampling: 100, size: 10, limit: 10_000, threshold: 4.3)
        .subscribe(on: DispatchQueue.global())
        .share()
    
    return working.eachWorkingBin(
        cancelableBags: &cancelableBags,
        internalCancelEmit: { b, _ in
            let now = Date().milisecondsSince1970
            let from = now - binHeaderDelayInMillis
            return SensorRecord.DropRecent(allocationNo: b.allocationNo, from: from, to: now, isFinished: true)
        },
        flowable: { b in
            detectControl(b.allocationNo)
                .removeDuplicates()
                .map({ it -> AnyPublisher<SensorRecord, Never> in
                    if it {
                        return sensorEvent.map({ eventDate -> SensorRecord in
                            return SensorRecord.Record(event: SensorDetectEvent(binHeader: b, location: Current.Shared.lastLocation, eventTime: eventDate))
                        })
                        .delaySubscription(for: .milliseconds(Int(workDelayInMillis)), scheduler: RunLoop.main)
                        .eraseToAnyPublisher()
                    } else {
                        let now = Date().milisecondsSince1970
                        let from = now - workDelayInMillis
                        return Just(SensorRecord.DropRecent(allocationNo: b.allocationNo, from: from, to: now, isFinished: false))
                            .eraseToAnyPublisher()
                    }
                })
                .switchToLatest()
                .delaySubscription(for: .milliseconds(Int(binHeaderDelayInMillis - max(0, -1*(b.startDate ?? 0)))), scheduler: RunLoop.main)
                .eraseToAnyPublisher()
        }).map({ $0.1 }).eraseToAnyPublisher()
}

extension AnyPublisher<[BinHeader], Never> {
    func eachWorkingBin<T: Any>(
        cancelableBags: inout Set<AnyCancellable>,
        internalCancelEmit: ((_ bin: BinHeader, _ lastElement: T?) -> T?)? = nil,
        flowable: @escaping (_ bin: BinHeader) -> AnyPublisher<T, Never>
    ) -> AnyPublisher<(BinHeader, T), Never> {
        let emitter = PassthroughSubject<(BinHeader, T), Never>()
        var onCancelEmit = internalCancelEmit
        let mapDisposable: MapDisposable<String> = MapDisposable()

        let d = self
        .receive(on: RunLoop.main)
        .sink(receiveValue: { ls in
            var m: [String : Cancellable] = [:]
            ls.filter({ $0.binStatus == "1"}).forEach({ b in
                let k = b.allocationNo
                let d = mapDisposable.take(key: k)
                
                var x: Cancellable? = nil
                if (d == nil || mapDisposable.isDisposed()) {
                    let consumer = { it in
                        emitter.send((b, it))
                    }
                    
                    x = flowable(b).withLastElementOnCancel(onCancel: { it in
                        if let rs = onCancelEmit?(b, it) {
                            consumer(rs)
                        }
                    }).sink(receiveValue: consumer)
                } else {
                    x = d
                }
                
                if let temp = m[k] {
                    temp.cancel()
                }
                m[k] = x
            })
            
            mapDisposable.clear()
            m.forEach({ (key, value) in
                _ = mapDisposable.put(key: key, d: value)
            })
        })
        
        cancelableBags.insert(d)
        
        emitter.handleEvents(receiveCancel:
                                 {
                                    onCancelEmit = nil
                                    d.cancel()
                                    mapDisposable.cancel()
                                }
                             )
        return emitter
            .eraseToAnyPublisher()
    }
}
